openslides_backend/migrations/migration_handler.py

```diff
@@ -0,0 +1,118 @@
+                        raise RuntimeError("Invalid migration handler state")
+                else:
+                    if MigrationHandler.migrate_thread_stream:
+                        # Migration finished, but last output was not read yet. Close stream and return result
+                        output = self.close_migrate_thread_stream()
+                        return {
+                            "status": MigrationProgressState.MIGRATION_FINISHED,
+                            "output": output,
+                        }
+                    else:
+                        # Nothing to report
+                        return {
+                            "status": MigrationProgressState.NO_MIGRATION_RUNNING,
+                            "message": "No migration running!",
+                        }
+
+            if MigrationHandler.migration_running:
+                raise View400Exception(
+                    "Migration is running, only 'progress' command is allowed"
+                )
+
+            if MigrationHandler.migrate_thread_stream:
+                self.logger.warning(
+                    "Deleting unread migration output:\n"
+                    + self.close_migrate_thread_stream()
+                )
+
+            verbose = payload.get("verbose", False)
+            MigrationHandler.migrate_thread_stream = StringIO()
+            if command in ("migrate", "finalize"):
+                thread = Thread(
+                    target=self.execute_migrate_command, args=[command, verbose]
+                )
+                thread.start()
+                thread.join(THREAD_WAIT_TIME)
+                if thread.is_alive():
+                    # Migration still running. Report current progress and return
+                    return {
+                        "status": MigrationProgressState.MIGRATION_RUNNING,
+                        "output": MigrationHandler.migrate_thread_stream.getvalue(),
+                    }
+                else:
+                    # Migration already finished/had nothing to do. Close stream and return all output
+                    return {
+                        "status": MigrationProgressState.MIGRATION_FINISHED,
+                        "output": self.close_migrate_thread_stream(),
+                    }
+            else:
+                # short-running commands can just be executed directy
+                self.execute_migrate_command(command, verbose)
+                return {
+                    "output": self.close_migrate_thread_stream(),
+                }
+
+    def execute_migrate_command(self, command: str, verbose: bool) -> None:
+        MigrationHandler.migration_running = True
+        time.sleep(10)
+        handler = MigrationWrapper(verbose, self.write_line)
+        try:
+            handler.execute_command(command)
+        except MigrationException as e:
+            raise View400Exception(str(e))
+        MigrationHandler.migration_running = False
```

Comment:
Shouldn't you put this in a finally clause to disable migration_running-state?

Reply:
Yes, good catch. That should also solve @peb-adr's issue with the invalid state after an invalid command. Will implement it.